src/source/avro.rs

```diff
@@ -0,0 +1,85 @@
+use byteorder::{BigEndian, ReadBytesExt};
+use constants;
+use metric;
+use mio;
+use serde_avro;
+use source;
+use std::{io, sync};
+use std::io::Read;
+use thread;
+use util;
+
+/// Avro source state
+pub struct Avro {
+    server: source::TCP,
+}
+
+/// Monitors for inbound system and TCP events.
+fn stream_handler(
+    chans: util::Channel,
+    tags: &sync::Arc<metric::TagMap>,
+    poller: &mio::Poll,
+    stream: mio::net::TcpStream,
+) {
+    let mut reader = io::BufReader::new(stream);
+    loop {
+        let mut events = mio::Events::with_capacity(1024);
+        match poller.poll(&mut events, None) {
+            Err(e) => panic!(format!("Failed during poll {:?}", e)),
+            Ok(_num_events) => for event in events {
+                match event.token() {
+                    constants::SYSTEM => return,
+                    _stream_token => {
+                        handle_avro_payload(chans.clone(), tags, &mut reader);
+                    }
+                }
+            },
+        };
+    }
+}
+
+/// Pulls length prefixed (4 bytes, BE), avro encoded
+/// binaries off the wire and populates them in the configured
+/// channel.
+fn handle_avro_payload(
+    mut chans: util::Channel,
+    _tags: &sync::Arc<metric::TagMap>,
+    reader: &mut io::BufReader<mio::net::TcpStream>,
+) {
+    let mut buf = Vec::with_capacity(4000);
```

Comment:
This was a copy and paste from other sources. 

As the buffer is resized later, I didn't think it really mattered what value was used here.

Reply:
It doesn't look like you use the `buf` until we know the size it needs to be. Perhaps you can just read the payload size off, and then create the buf with the required capacity, then resize it and read into it, to prevent a potential reallocation.

Reply:
Solid point.

I moved the buffer creation to after reading the length off the wire.

Reply:
For this to avoid potential re-allocations the buf will have to live outside of `handle_avro_payload`. Right now we're allocating a buf of 4000 each time we enter the function, then resizing it.

Reply:
Ah, but you can't do that anyhow since that buffer is going to be owned by the event. Never mind.